### ISSUE TITLE
paywall proxies can make every domain seem reachable

### ISSUE THREAD

**Issue by dylang** [sindresorhus/is-reachable#3]
When on a network that has a paywall proxy, such as GoGo WiFi on a plane, `is-reachable` is true for basically any hostname because the paywall proxy always returns with their information page about paying for the service.

I don't know if this is fixable for port 80, but I think #1 will avoid the problem for use cases that aren't ports that get automatically redirected by paywall proxies.

**Comment by silverwind** [comment]
Good point, we haven't thought about captive portals yet. While it's probably right that most portals only do port 80 (can't easily do 443), I think we should try to make it work universally for all ports.

We need to somehow verify that we're connecting to the real target. I *think* most portals do their intercepting through DNS by returning a private IP address for all lookups, but I could be wrong.

It would be trivial to test if the target resolves to an private address, but that would break checking servers on the local network. Hmm.

**Comment by silverwind** [comment]
One solution I see could be to `dns.lookup` the hostname first. If it resolves to a private address but if the hostname itself is in the form of `domain.tld` (as opposed to `domain` which is common for intranets), return unreachability. It's not 100% proof for cases a public-looking domain name is used in intranets, but it's the best I could think of right now.

According to Wikipedia, a less common variant to implement captive portals is ICMP redirection. It *might* be possible to detect this too through DNS.

**Comment by silverwind** [comment]
To properly do it, we'd need a list of all current IANA-approved top level domains (https://data.iana.org/TLD/tlds-alpha-by-domain.txt), which I could put into a module and maintain. There would be false-negatives if someone decides to host an official top level domain in private IP space, but I think these cases will likely be almost nil.

**Comment by kevva** [comment]
https://www.npmjs.com/package/tlds

**Comment by silverwind** [comment]
:+1: exactly what I need. I have to brush up my npm search skills.

**Comment by silverwind** [comment]
@dylang I'd really appreciate if you could test the current git version on such a WiFi. It should now handle DNS redirection (which I assume is what you see).

Myself, I will try setting up a guest WiFi with a captive portal to test on my home router. Also ICMP redirection should be tested, I'll try to set that up on my router too, no promises thought.

**Comment by sindresorhus** [comment]
ping @dylang

**Comment by dylang** [comment]
Hi @sindresorhus and @silverwind - thanks for trying to address this issue. 

Based on Github master, it has not been resolved:

![screen shot 2015-09-08 at 11 39 25 am](https://cloud.githubusercontent.com/assets/51505/9739587/7b57bcd8-561e-11e5-88c2-3236fa03dac3.png)
*you may need to click on that image to read it*

You'll see `//switched internet` when I switched to a router that requires login. The google ping doesn't response, but we still get `true` back from `is-reachable`.

**Comment by silverwind** [comment]
@dylang could you post the output of `dig google.com` and `nc -v {ip} 80` (`ip being the ip obtained from dig`) when internet is switched?

**Comment by dylang** [comment]
@silverwind Sure thing!

![screen shot 2015-09-08 at 12 35 37 pm](https://cloud.githubusercontent.com/assets/51505/9741059/3ec01e84-5626-11e5-8083-f61fd912bf08.png)

I tried a couple different google.com ip's and got the same results.

**Comment by silverwind** [comment]
@dylang So that proxy blocks ping which is fine. DNS still seems to work as expected. Are you're getting their paywall now when surfing to google?

**Comment by silverwind** [comment]
Oh, and because I suspect there could be ICMP redirection involved. Could you possibly post `traceroute 173.194.121.41` one time with and without the proxy?

**Comment by dylang** [comment]
Going to `google.com` on port 80 in my browser forwards me to `https://1.1.1.1/login.html?redirect=google.com`.

![screen shot 2015-09-08 at 2 41 38 pm](https://cloud.githubusercontent.com/assets/51505/9744091/d0fc8c22-5637-11e5-9304-a9e9d3ec1fac.png)

**Comment by silverwind** [comment]
Thanks. One last request: `curl -v google.com`

**Comment by dylang** [comment]
@silverwind enjoying helping you with this, wish I could be faster!

<img width="908" alt="screen shot 2015-09-08 at 8 00 03 pm" src="https://cloud.githubusercontent.com/assets/51505/9750350/82be121a-5664-11e5-9a38-5a36a7de7692.png">

**Comment by silverwind** [comment]
Interesting. That's certainly not the kind of proxy I was expecting. It looks like it's answering on behalf of every IP. Is see no clean way of detecting that yet.

If you're still interested, could you obtain a Wireshark packet capture in `.pcapng` format of this `curl` request and mail me that? I only need the request and response packet, and possibly any ICMP that happens in between them.

Use one of these Wireshark builds: https://www.wireshark.org/download/automated/osx/

**Comment by sindresorhus** [comment]
I'm also behind a login wall and `is-reachable` gives me a false positive.

When I go to `google.com` I get redirected to `http://192.168.2.254/portal/user-authen.php?reqUrl=http%3A%2F%2Fgoogle.com`.


```
❯ dig google.com

; <<>> DiG 9.8.3-P1 <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 30673
;; flags: qr rd ra; QUERY: 1, ANSWER: 16, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		7	IN	A	110.164.10.50
google.com.		7	IN	A	110.164.10.30
google.com.		7	IN	A	110.164.10.29
google.com.		7	IN	A	110.164.10.24
google.com.		7	IN	A	110.164.10.35
google.com.		7	IN	A	110.164.10.40
google.com.		7	IN	A	110.164.10.45
google.com.		7	IN	A	110.164.10.44
google.com.		7	IN	A	110.164.10.49
google.com.		7	IN	A	110.164.10.34
google.com.		7	IN	A	110.164.10.20
google.com.		7	IN	A	110.164.10.25
google.com.		7	IN	A	110.164.10.59
google.com.		7	IN	A	110.164.10.39
google.com.		7	IN	A	110.164.10.54
google.com.		7	IN	A	110.164.10.55

;; Query time: 8 msec
;; SERVER: 192.168.2.254#53(192.168.2.254)
;; WHEN: Wed Sep  9 12:41:27 2015
;; MSG SIZE  rcvd: 284
```

```
❯ nc -v 110.164.10.50 80
found 0 associations
found 1 connections:
     1:	flags=82<CONNECTED,PREFERRED>
	outif en0
	src 192.168.2.31 port 58816
	dst 110.164.10.50 port 80
	rank info not available
	TCP aux info available

Connection to 110.164.10.50 port 80 [tcp/http] succeeded!
```

```
traceroute to 173.194.121.41 (173.194.121.41), 64 hops max, 52 byte packets
 1  login (192.168.2.254)  1.885 ms  1.406 ms  1.408 ms
 2  * * *
 3  * * *
 4  * * *
 5  * * *
 6  * * *
 7  * * *
 8  * * *
 9  * * *
10  * * *
```

And traceroute when logged into the login wall:

```
❯ traceroute 173.194.121.41
traceroute to 173.194.121.41 (173.194.121.41), 64 hops max, 52 byte packets
 1  login (192.168.2.254)  1.323 ms  1.252 ms  1.152 ms
 2  mx-ll-110.164.0-155.static.3bb.co.th (110.164.0.155)  57.219 ms
    mx-ll-110.164.0-153.static.3bb.co.th (110.164.0.153)  27.018 ms
    mx-ll-110.164.0-155.static.3bb.co.th (110.164.0.155)  27.299 ms
 3  mx-ll-110.164.0-152.static.3bb.co.th (110.164.0.152)  26.447 ms  26.493 ms  28.826 ms
 4  mx-ll-110.164.1-121.static.3bb.co.th (110.164.1.121)  25.595 ms  325.133 ms  43.365 ms
 5  * * *
 6  mx-ll-110.164.14-237.static.3bb.co.th (110.164.14.237)  51.298 ms  49.655 ms
    mx-ll-110.164.14-235.static.3bb.co.th (110.164.14.235)  49.465 ms
 7  72.14.195.182 (72.14.195.182)  63.493 ms  102.187 ms *
 8  209.85.242.246 (209.85.242.246)  118.981 ms
    209.85.242.236 (209.85.242.236)  171.285 ms
    209.85.242.246 (209.85.242.246)  131.127 ms
 9  209.85.242.232 (209.85.242.232)  315.393 ms * *
10  209.85.255.131 (209.85.255.131)  302.335 ms  125.104 ms
    209.85.255.129 (209.85.255.129)  121.198 ms
11  * 209.85.143.3 (209.85.143.3)  292.854 ms  218.371 ms
12  * 72.14.239.208 (72.14.239.208)  264.299 ms  248.518 ms
13  209.85.143.214 (209.85.143.214)  301.566 ms
    209.85.143.114 (209.85.143.114)  303.714 ms  284.251 ms
14  * * *
15  72.14.233.97 (72.14.233.97)  274.031 ms  314.182 ms  312.364 ms
16  iad23s26-in-f9.1e100.net (173.194.121.41)  267.944 ms  284.953 ms  328.379 ms
```

-

Before login to login wall:

```
❯ curl -v google.com       
* Rebuilt URL to: google.com/
*   Trying 110.164.10.55...
* Connected to google.com (110.164.10.55) port 80 (#0)
> GET / HTTP/1.1
> Host: google.com
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer: 
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 302 Moved Temporarily
< Date: Wed, 09 Sep 2015 05:48:15 GMT
< Server: Apache
< X-Powered-By: PHP/5.3.2
< Location: http://192.168.2.254/portal/user-authen.php?reqUrl=http%3A%2F%2Fgoogle.com
< Content-Length: 484
< Connection: close
< Content-Type: text/html
< 
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"> 
<!--
	iBSG-Lite GUI
	Copyright (C) N.V.K.INTER CO., LTD
	-- Html header --
-->
<head>
<meta http-equiv="pragma" content="no-cache">
</head>
<body>Browser does not support redirects!, please click <a href="http://192.168.2.254/portal/user-authen.php?reqUrl=http%3A%2F%2Fgoogle.com">here</a></body>
</html>
* Closing connection 0
```

After login to login wall:

```
❯ curl -v google.com
* Rebuilt URL to: google.com/
*   Trying 110.164.10.55...
* Connected to google.com (110.164.10.55) port 80 (#0)
> GET / HTTP/1.1
> Host: google.com
> User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
> Accept: */*
> Referer: 
> 
< HTTP/1.1 302 Found
< Location: http://www.google.co.th/?gws_rd=cr&ei=SMjvVcS8IY-gugTq0JrgDg
< Cache-Control: private
< Content-Type: text/html; charset=UTF-8
< P3P: CP="This is not a P3P policy! See http://www.google.com/support/accounts/bin/answer.py?hl=en&answer=151657 for more info."
< Date: Wed, 09 Sep 2015 05:48:56 GMT
< Server: gws
< Content-Length: 261
< X-XSS-Protection: 1; mode=block
< X-Frame-Options: SAMEORIGIN
< Set-Cookie: PREF=ID=1111111111111111:FF=0:TM=1441777736:LM=1441777736:V=1:S=V1GRv_qcXaZxE8Uk; expires=Thu, 31-Dec-2015 16:02:17 GMT; path=/; domain=.google.com
< Set-Cookie: NID=71=iUiapRGZ5BG3_TBoKN5mwj6_UG_ukHBrhPns2u_MnGb20SK-ktWq4K69Fh-Rokl-EnWdknx9Us1b7MvQHAHhmJwbZFpPGBlX8KpxCK2CmevEOTmPdi7X_PuBZs4JhOf3; expires=Thu, 10-Mar-2016 05:48:56 GMT; path=/; domain=.google.com; HttpOnly
< 
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="http://www.google.co.th/?gws_rd=cr&amp;ei=SMjvVcS8IY-gugTq0JrgDg">here</A>.
</BODY></HTML>
* Connection #0 to host google.com left intact
```

**Comment by silverwind** [comment]
@sindresorhus At least in your case, the IP you're redirected to is a private one. We could add some special checks the Location header in case Port  80. For `443`, I assume a `curl -v https://google.com` does not work, correct?

The `Location: https://1.1.1.1` nonsense in @dylang's case is a bit infuriating as that IP could very well be in use. At least the misuse of the `1.0.0.0/8` range is [common](http://www.potaroo.net/studies/1slash8/1slash8.html) and appears to mostly be done by Cisco Wireless gateways.

@dylang thanks for all info, I don't think I really need that packet capture any more. I'll think about a solution now :)

**Comment by sindresorhus** [comment]
Yup:

```
❯ curl -v https://google.com
* Rebuilt URL to: https://google.com/
*   Trying 110.164.10.59...
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
*   Trying 2404:6800:4001:807::200e...
* Immediate connect fail for 2404:6800:4001:807::200e: No route to host
```

**Comment by silverwind** [comment]
Ah, that's at least something, so one can use `google.com:443` for a more reliable check. I'll work on getting port 80 fixed though.

**Comment by silverwind** [comment]
@dylang In case you ever come in contact again with that proxy, I'd love if you could test against `master`, you case should be handled now.

**Comment by dylang** [comment]
Sounds great @silverwind. I don't have any travel scheduled, but I'll be sure to try it next time I am in the air or in a hotel.

**Comment by silverwind** [comment]
Latest changes are still not released. If someone is a owner or has access to Cisco Access Points that feature a login page, please test latest master so we can be sure the changes work!

**Comment by sindresorhus** [comment]
@silverwind Maybe time to do a new release? Doesn't look like we're going to get more feedback on this in the short-term.

**Comment by silverwind** [comment]
Okay, released 1.3.0.
